action.php

```diff
@@ -91,7 +91,8 @@
             if ($data = $form->get_data()) {
                 if ($factorobject->setup_user_factor($data)) {
                     $factorobject->set_state(\tool_mfa\plugininfo\factor::STATE_PASS);
-                    redirect($returnurl);
+                    $finalurl = new moodle_url($returnurl, array('setup' => $factorobject->name));
```

Comment:
This whole thing can be much simpler, there is api helpers for doing this as part of a redirect, or sticking it into the session and core handles it:

https://docs.moodle.org/dev/Notifications#Adding_notifications